pandas/tests/frame/test_alter_axes.py

```diff
@@ -1294,7 +1294,7 @@ def test_rename_mapper_multi(self):
     def test_rename_positional_named(self):
         # https://github.com/pandas-dev/pandas/issues/12392
         df = DataFrame({"a": [1, 2], "b": [1, 2]}, index=["X", "Y"])
-        result = df.rename(str.lower, columns=str.upper)
+        result = df.rename(index=str.lower, columns=str.upper)
```

Comment:
So did this break? If so, is it solely because of the keyword only change to `index` and `columns`?

If so, I might prefer a proper deprecation warning for this change... But could be convinced otherwise, since I'd consider the old code bad style.

Reply:
This would be related to https://github.com/pandas-dev/pandas/pull/29140/files#r337242915

This is a little tough to support without keeping the ambiguity, and it doesn't really work outside of this one niche application.

For instance, if you tried to apply `str.lower` to the columns and `str.upper` to the index you would get the following error:

```python
>>> df.rename(str.lower, index=str.upper, axis=1)
TypeError: Cannot specify both 'axis' and any of 'index' or 'columns'.
```

Even explicitly stating the axis in this case would raise an error:

```python
>>> df.rename(str.lower, columns=str.upper, axis=0)
TypeError: Cannot specify both 'axis' and any of 'index' or 'columns'.
```

So we could *still* support this but it would take some gymnastics and not sure it makes things clearer, though open to the thoughts of others.

Maybe can be clarified further in the whatsnew to mitigate confusion?